README.md
# pdf2canvas

[![NPM version](https://img.shields.io/npm/v/pdf2canvas.svg)](https://www.npmjs.com/package/pdf2canvas)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](/LICENSE)

Convert a pdf to image dataURL / file

## Installation

```sh
$ npm install pdf2canvas
```

## Initialize

### Example

```js
const Pdf2Canvas = require('pdf2canvas');

const pdfCanvas = new Pdf2Canvas('./test.pdf');
```

## API

### .downloadPNG (options)

- Convert a pdf to PNG file(s) asynchronously
- `options`: {Object}
  - `pageRange`: {Array{number}}
    - the start page number and the last page number
    - default to `[1, Infinity]`
  - `outputDir`: {string}
    - the specified directory to save image files
    - default to `./`
  - `viewportScale`: {number} | {Function}
    - the viewport scale
    - in case of a function is specified, the original width & height will be passed in as its arguments
    - default to `1.5`
  - `config`: {Object}
    - PNG encode config
    - `compressionLevel`: ZLIB compression level (between 0 and 9), default to 6
    - `filters`: the compression filter(s), default to `PNG-ALL-FILTERS`
    - `palette`: the palette (indexed PNGs only), default to undefined
    - `backgroundIndex`: the background palette index (indexed PNGs only), default to 0
    - `resolution`: the resolution in pixels per meter (ppi), default to undefined
- Returns: {Promise} containing the saved file paths {Array{string}}

```js
const result = await pdfCanvas.downloadPNG({
  outputDir: './images',
  pageRange: [1, 2],
  viewportScale: (width, height) => {
    return width > height ? 2200 / width : 1600 / width;
  },
});
// => ["images/page-1.png", "images/page-2.png"]
```

### .downloadJPEG (options)

- Convert a pdf to JPEG file(s) asynchronously
- `options`: {Object}
  - `pageRange`: {Array{number}}
    - the start page number and the last page number
    - default to `[1, Infinity]`
  - `outputDir`: {string}
    - the specified directory to save image files
    - default to `./`
  - `viewportScale`: {number} | {Function}
    - the viewport scale
    - in case of a function is specified, the original width & height will be passed in as its arguments
    - default to `1.5`
  - `config`: {Object}
    - JPEG encode config
    - `quality`: the quality (0 to 1), default to 0.8
    - `progressive`: if progressive compression should be used, default to false
    - `chromaSubsampling`: if chroma subsampling should be used, default to true
- Returns: {Promise} containing the saved file paths {Array{string}}

```js
const result = await pdfCanvas.downloadJPEG({
  outputDir: './images',
  pageRange: 1,
  viewportScale: 2,
  config: { quality: 0.9 },
});
// => ["images/page-1.jpg", "images/page-2.jpg"]
```

### .toDataURL (options)

- Convert a pdf to PNG / JPEG dataURL asynchronously
- `options`: {Object}
  - `pageRange`: {Array{number}}
    - the start page number and the last page number
    - default to `[1, Infinity]`
  - `viewportScale`: {number} | {Function}
    - the viewport scale
    - in case of a function is specified, the original width & height will be passed in as its arguments
    - default to `1.5`
  - `isPNG`: {boolean}
    - PNG or JPEG
    - default to true
  - `quality`: {number}, optional
    - the quality (0 to 1)
    - default to 0.8
- Returns: {Promise} containing dataURLs {Array{string}}

```js
const result = await pdfCanvas.toDataURL({
  pageRange: [1, 3],
  viewportScale: 1.2,
  isPNG: false,
  quality: 0.9,
});
// => ["data:image/jpeg;base64,...", ...]
```

### [MIT Licensed](LICENSE)


HISTORY.md
0.1.0 / 2020-04-10
==================

  * Initial Publish


NodeCanvasFactory.js
// https://github.com/mozilla/pdf.js/blob/cb01a7de09516cf9d91602ec2892609699b03438/examples/node/pdf2png/pdf2png.js#L20
var Canvas = require('canvas');
var assert = require('assert').strict;

function NodeCanvasFactory() {}
NodeCanvasFactory.prototype = {
  create: function NodeCanvasFactory_create(width, height) {
    assert(width > 0 && height > 0, 'Invalid canvas size');
    var canvas = Canvas.createCanvas(width, height);
    var context = canvas.getContext('2d');
    return {
      canvas: canvas,
      context: context,
    };
  },

  reset: function NodeCanvasFactory_reset(canvasAndContext, width, height) {
    assert(canvasAndContext.canvas, 'Canvas is not specified');
    assert(width > 0 && height > 0, 'Invalid canvas size');
    canvasAndContext.canvas.width = width;
    canvasAndContext.canvas.height = height;
  },

  destroy: function NodeCanvasFactory_destroy(canvasAndContext) {
    assert(canvasAndContext.canvas, 'Canvas is not specified');

    // Zeroing the width and height cause Firefox to release graphics
    // resources immediately, which can greatly reduce memory consumption.
    canvasAndContext.canvas.width = 0;
    canvasAndContext.canvas.height = 0;
    canvasAndContext.canvas = null;
    canvasAndContext.context = null;
  },
};

module.exports = NodeCanvasFactory;


index.js
'use strict';

const fs = require('fs');
const path = require('path');
const PDFJSLib = require('pdfjs-dist');
const { createCanvas } = require('canvas');
const NodeCanvasFactory = require('./NodeCanvasFactory');

const isArray = Array.isArray;
const isBuffer = Buffer.isBuffer;

const isNumber = (value) => typeof value === 'number';
const isFunction = (value) => typeof value === 'function';

const defaultPageRange = [1, Infinity];
const defaultJpegQuality = 0.8;

class Pdf2Canvas {
  constructor(pdfPathOrBuffer, { viewportScale } = {}) {
    if (isBuffer(pdfPathOrBuffer)) {
      this.pdfPath = null;
      this.pdfBuffer = pdfPathOrBuffer;
    } else {
      this.pdfPath = pdfPathOrBuffer;
      this.pdfBuffer = fs.readFileSync(pdfPathOrBuffer);
    }

    this.rowData = new Uint8Array(this.pdfBuffer);
    this.viewportScale = viewportScale;
  }

  async toDataURL({ pageRange, viewportScale, isPNG, quality } = {}) {
    return this.convert({ isDataURL: true, pageRange, viewportScale, isPNG, config: quality });
  }

  async download({ pageRange, outputDir, viewportScale, isPNG, config } = {}) {
    return this.convert({ isDataURL: false, pageRange, outputDir, viewportScale, isPNG, config });
  }

  async downloadPNG({ pageRange, outputDir, viewportScale, config } = {}) {
    return this.convert({ isDataURL: false, pageRange, outputDir, viewportScale, isPNG: true, config });
  }

  async downloadJPEG({ pageRange, outputDir, viewportScale, config } = {}) {
    return this.convert({ isDataURL: false, pageRange, outputDir, viewportScale, isPNG: false, config });
  }

  async convert({
    pageRange = defaultPageRange,
    viewportScale = this.viewportScale || 1.5,
    outputDir = './',
    isPNG = true,
    isDataURL = false,
    config,
  } = {}) {
    if (isArray(pageRange)) {
      pageRange.sort();
      if (pageRange[0] < 1) pageRange = defaultPageRange;
    } else if (isNumber(pageRange) && pageRange > 0) {
      pageRange = [pageRange, pageRange];
    } else {
      pageRange = defaultPageRange;
    }

    try {
      const doc = await PDFJSLib.getDocument({
        data: this.rowData,
        nativeImageDecoderSupport: PDFJSLib.NativeImageDecoding.DISPLAY,
      }).promise;

      const numPages = doc.numPages;
      const minPage = pageRange[0];
      const maxPage = pageRange[1] > numPages ? numPages : pageRange[1];

      const results = [];

      for (let x = minPage; x <= maxPage; x++) {
        const page = await doc.getPage(x);

        let viewport;
        if (isNumber(viewportScale)) {
          viewport = page.getViewport({ scale: viewportScale });
        } else if (isFunction(viewportScale)) {
          viewport = page.getViewport({ scale: 1 });
          const scale = viewportScale(viewport.width, viewport.height);
          viewport = page.getViewport({ scale });
        }

        const canvas = createCanvas(viewport.width, viewport.height);
        const context = canvas.getContext('2d');
        const canvasFactory = new NodeCanvasFactory();

        const renderContext = {
          canvasContext: context,
          viewport,
          canvasFactory,
        };

        await page.render(renderContext).promise;

        if (isDataURL) {
          const dataURL = canvas.toDataURL(isPNG ? 'image/png' : 'image/jpeg', config || defaultJpegQuality);
          results.push(dataURL);
        } else {
          const { stream, ext } = isPNG ? this.createPNGStream(canvas, config) : this.createJPEGStream(canvas, config);

          const filePath = path.join(outputDir, `page-${x}.${ext}`);
          await this.writeStreamToFile(stream, filePath);
          results.push(filePath);
        }
      }

      return results;
    } catch (err) {
      console.error(err);
      return Promise.reject(err);
    }
  }

  createPNGStream(
    canvas,
    {
      compressionLevel = 6,
      filters = canvas.PNG_ALL_FILTERS,
      palette = undefined,
      backgroundIndex = 0,
      resolution = undefined,
    } = {}
  ) {
    const stream = canvas.createPNGStream({
      compressionLevel,
      filters,
      palette,
      backgroundIndex,
      resolution,
    });

    return { stream, ext: 'png' };
  }

  createJPEGStream(canvas, { quality = defaultJpegQuality, progressive = false, chromaSubsampling = true } = {}) {
    const stream = canvas.createJPEGStream({
      quality,
      progressive,
      chromaSubsampling,
    });

    return { stream, ext: 'jpg' };
  }

  writeStreamToFile(readableStream, filePath) {
    let writableStream = fs.createWriteStream(filePath);

    readableStream.pipe(writableStream);

    return new Promise(function (resolve, reject) {
      readableStream.once('error', reject);
      writableStream.once('error', reject);
      writableStream.once('finish', resolve);
    }).catch(function (err) {
      readableStream = null; // Explicitly null because of v8 bug 6512.
      writableStream.end();
      throw err;
    });
  }
}

module.exports = Pdf2Canvas;
